Add validate utility (#1820)
strapi/documentation: docusaurus/docs/dev-docs/backend-customization/controllers.md

When a new [content-type](/dev-docs/backend-customization/models#content-types) is created, Strapi builds a generic controller with placeholder code, ready to be customized.
:::

### Sanitization and Validation in controllers

:::warning

> It's strongly recommended you sanitize (v4.8.0) and/or validate (v4.13.0) your incoming request query and parameters utilizing the new `sanitizeQuery` and `validateQuery` functions to prevent leaking of private data.

Reviewer: ```suggestion
It's strongly recommended you sanitize (v4.8.0+) and/or validate (v4.13.0+) your incoming request query and parameters utilizing the new `sanitizeQuery` and `validateQuery` functions to prevent the leaking of private data.
```

I added "+" because I assumed you meant "with these versions and onwards". Do you know if my assumption was correct?

> It's strongly recommended you sanitize (v4.8.0+) and/or validate (v4.13.0+) your incoming request query and parameters utilizing the new `sanitizeQuery` and `validateQuery` functions to prevent the leaking of private data.

Author: ```suggestion
It's strongly recommended you sanitize (v4.8.0+) and/or validate (v4.13.0+) your incoming request query utilizing the new `sanitizeQuery` and `validateQuery` functions to prevent the leaking of private data.
```

:::

#### Sanitization when utilizing controller factories

Within the Strapi factories the following functions are exposed that can be used for sanitization and validation:

| Function Name    | Parameters                 | Description                                                                          |
|------------------|----------------------------|--------------------------------------------------------------------------------------|
| `sanitizeQuery`  | `ctx`                      | Sanitizes the request query                                                          |
| `sanitizeOutput` | `entity`/`entities`, `ctx` | Sanitizes the output data where entity/entities should be an object or array of data |
| `sanitizeInput`  | `data`, `ctx`              | Sanitizes the input data                                                             |
| `validateQuery`  | `ctx`                      | Validates the request query (throws an error on invalid params)                      |

> | `validateInput`  | `data`, `ctx`              | Validates the input data (throws an error on invalid data)                           |

Author: ```suggestion
| `validateInput`  | `data`, `ctx`              | (EXPERIMENTAL) Validates the input data (throws an error on invalid data)                           |
```

These functions automatically inherit the sanitization settings from the model and sanitize the data accordingly based on the content-type schema and any of the content API authentication strategies, such as the Users & Permissions plugin or API tokens.

:::warning

> Because these methods use the model associated with the current controller, if you query data that is from another model (ie, doing a find for "menus" within a "restaurant" controller method) you must instead use the `@strapi/utils` tools, such as `sanitize.contentAPI.query` described below, or else the result of your query will be sanitized against the wrong model.

Reviewer: ```suggestion
Because these methods use the model associated with the current controller, if you query data that is from another model (i.e., doing a find for "menus" within a "restaurant" controller method), you must instead use the `@strapi/utils` tools, such as `sanitize.contentAPI.query` described below, or else the result of your query will be sanitized against the wrong model.
```

</TabItem>
</Tabs>

#### Sanitization and validation when building custom controllers

Within custom controllers, there are 5 primary functions exposed via the `@strapi/utils` package that can be used for sanitization and validation:

| Function Name                | Parameters         | Description                                             |
|------------------------------|--------------------|---------------------------------------------------------|
| `sanitize.contentAPI.input`  | `data`, `schema`, `auth`      | Sanitizes the request input including non-writable fields, removing restricted relations, and other nested "visitors" added by plugins |
| `sanitize.contentAPI.output` | `data`, `schema`, `auth`      | Sanitizes the response output including restricted relations, private fields, passwords, and other nested "visitors" added by plugins  |
| `sanitize.contentAPI.query`  | `ctx.query`, `schema`, `auth` | Sanitizes the request query including filters, sort, fields, and populate  |
| `validate.contentAPI.query`  | `ctx.query`, `schema`, `auth` | Validates the request query including filters, sort, fields (currently not populate) |

> | `validate.contentAPI.input`  | `data`, `schema`, `auth` | Validates the request input including non-writable fields, removing restricted relations, and other nested "visitors" added by plugins |

Author: ```suggestion
| `validate.contentAPI.input`  | `data`, `schema`, `auth` | (EXPERIMENTAL) Validates the request input including non-writable fields, removing restricted relations, and other nested "visitors" added by plugins |
```
